src/mjolnir/pbfgraphparser.cc

```diff
@@ -374,19 +365,32 @@ struct graph_callback : public OSMPBF::Callback {
     // Add the refs to the reference list and mark the nodes that care about when processing nodes
     loop_nodes_.clear();
     for (size_t i = 0; i < nodes.size(); ++i) {
+      // If we've seen this node anywhere else then its an intersection (ie graph node)
       const auto& node = nodes[i];
       if (shape_.get(node)) {
         intersection_.set(node);
         ++osmdata_.edge_count;
       } else {
         ++osmdata_.node_count;
       }
-      way_nodes_->push_back({{node}, static_cast<uint32_t>(ways_->size()), static_cast<uint32_t>(i)});
       shape_.set(node);
+
+      // Check whether the node is on a part of a way doubling back on itself
+      OSMNode osm_node{node};
+      auto inserted = loop_nodes_.insert(std::make_pair(node, i));
+      bool flattening = inserted.first->second > 0 && i < nodes.size() - 1 &&
+                        nodes[i + 1] == nodes[inserted.first->second - 1];
+      bool unflattening = i > 0 && inserted.first->second < nodes.size() - 1 &&
+                          nodes[i - 1] == nodes[inserted.first->second + 1];
+      osm_node.flat_loop_ = flattening || unflattening;
```

Comment:
previously this got triggered whenever we found the point at which at double back occured or when a node was not unique. it turns out we need to allow nodes to be not unique in the case of legitimate self intersections. so instead we harden to just the case where the going and outgoing paths at a non unique node are the same (ie is a flat loop). the cool thing is that this work also for the node where the double back occurs even though its not duplicated it has the same ingoing and outgoing path. in this case path means neighboring node ids.